sympy/sets/sets.py

```diff
@@ -994,6 +993,29 @@ def _eval_imageset(self, f):
             return
         var = f.variables[0]
 
+        if expr.is_Piecewise:
+            result = S.EmptySet
+            domain_set = self
+            for i in expr.args:
+                func = i[0]
+                if i[1] is S.true:
+                    intrvl = domain_set
+                else:
+                    intrvl = i[1].as_set()
+                    intrvl = Intersection(domain_set, intrvl)
+
+                if func.is_Number:
+                    image = FiniteSet(func)
+                else:
+                    image = imageset(Lambda(var, func), intrvl)
+                result = Union(result, image)
+
+                # remove the part which has been `imaged`
+                domain_set = Complement(domain_set, intrvl)
+                if domain_set.is_EmptySet:
+                    break
```

Comment:
> I'm not sure how does line 1012-1015 helps

They help since, if the `self( or domain_set)` is exhausted, then there is no point in considering how the `piecewise function` has been defined. Ex. `imageset(x, Piecewise((x, x <= 2), (x**2, True)), Interval(0, 2))` , if i want to find the `imageset` in this over `Interval(0, 2)` now there is no point in considering `how the piecewise function` has been defined for `x > 2` (it does not matter how it is defined) that is taken care of by the `domain_set.is_EmptySet` i.e. as the domain_set is `exhausted`, the `loop breaks`.

> Are there cases when this statement will make the loop break before it supposed to break?

In the above example this happens. The test cases included such.